### DISCUSSION TITLE
edits we created for iw3 in our discord server

### DISCUSSION THREAD

**Discussion by loawizard** [nagadomi/nunif#80] [Ideas]
Hello!

First of all, thank you so much for making this amazing software available for everyone!
It's truly amazing.

me and a group of 2D to 3D fanatics have created some edits to the gui and utils files to unlock it even more!
We have created a discord where we test out Iw3 on different PC setups, and we would be honoured if you could join.

We want to help you improve Iw3 so it becomes the best 2D to 3D converter ever! Also, I can make you the co-owner of the Discord, as it is your software!

Here is the link to the discord : https://discord.gg/T72KQJ6z

Thank you so much, and I hope to talk to you soon.

**Comment by nagadomi** [comment]
I will respond to bug reports, questions, and feature requests posted to this repository, but I basically won't cooperate in any other way.

I am not saying that I dislike forked software or external communities.
Feel free to use the source code under the license.

**Reply by gituser123456789000** [reply]
You really did excellent work. This is now my go-to for 2d to 3d conversions.

We unlocked settings limitations, because with the program we came from, it was ideal to use negative Strength settings, so we wanted to try that. We also unlocked the limitations for Convergence Plane and added larger Edge Fix values, again, for testing purposes, but your default settings, or near them, seem to give the best results anyway.

A couple real improvements we did make though, and I hope you'll add these to the main program.. for Depth Resolution, we added a bunch of more values to make better quality (higher resolution depth maps). I read what computers you tested the program on and I see why you maxed it out at 512, which works with your 8GB VRAM and the Default worked with 4GB VRAM laptop, but it limits the quality of what higher end PCs are capable of. Using a 4090 and having 64GB of regular RAM, that gives me 24GB (VRAM) + 32GB (Shared RAM).. 56GB of total GPU RAM. This lets me consistently use 1024 Depth Resolution and I was even able to use 1536 Depth Resolution for certain content. Your program can certainly benefit from adding higher end Depth Resolution options, and most ideally, allow there to be a custom input (same as the 3D Strength and Convergence Plane drop-down boxes have), so users can pick any value. This will cover all levels of GPU and RAM totals, so everyone can get the best possible results for their hardware, and it also makes the program more future-proof for when more VRAM and regular RAM combinations come out.. like at least 256GB RAM (at least 128GB would be Shared) is already possible with DDR5. We added drop-down values of 640, 720, 768, 832, 1024, 1536, 2048 to cover more people's capabilities, but a custom value option would again be most ideal.

Lastly, we changed the file name output ending from "_LRF" to "LR_Flat" so the files automatically should open as flat LR (SBS) videos and photos immediately in VR players, instead of possibly opening as whatever your most recent settings may have been, like VR180.. then you have to start changing every photo or video you view to display properly.

Those last 2 are good and easy upgrades to the program that I think you could implement quickly to the main program, as we aren't coders at all, but were able to do it.

Additionally, people have requested the ability to resume failed conversions, like if they suffer a crash, they wish they could go back in and continue from where they failed. Adding the ability to produce Anaglyph and Cross-eyed are other ideas.. just as extra abilities, not too many people are dying without them, it would just be nice to have everything in one program. Another idea was the ability to use your own depth maps. Some people can, or think they can, create or enhance depth maps further than some single programs, so they would like the ability to use their own custom depth maps in the conversion process. These are obviously more involved requests, but custom Depth Resolution values would be a great next update.

You did a fantastic job with this program. Thank you nagadomi.

**Reply by nagadomi** [reply]
Thank you for the feedback.
This is a quick reply.

> settings limitations

I will try to unlock GUI limits if possible. CLI does not have some settings limitations.

>Negative Strength

I think negative 3D Strength is same feature as L/R Swap. VR video players probably have this feature. It is better to use that.

Also, note that `3D Strength` and `Convergence Plane` values are used as input to the machine learning model(AI), not as mathematical formula.
That model has only trained for `3D Strength: 0-2.5` and `Convergence Plane: 0.0-1.0`, so outside of that range behavior is undefined. I know it works fine in practice, though.
Since this limitation comes from the old version of the implementation, it is possible to re-train with more wider range of values.

>Depth Resolution

I will unlock the input limitations on GUI.

However, the depth resolution setting has more complex problem other than VRAM usage.
Increasing the resolution of the depth will cause bad results at a certain point.
This is due to the image resolution used when training the model and its parameter size.

ZoeDepth model is trained with height=384px images. DepthAnything model is trained with 518x518 images. I think about 2x that resolution is the upper limit for working properly. (not tested)

To output a higher resolution depth, an approach such as [BoostingMonocularDepth](https://github.com/compphoto/BoostingMonocularDepth) should be used.
However, BoostingMonocularDepth has the problem of license and it is very slow.

>LR_Flat

This is dependent on the video player. Which software requires this?
For now, supports
https://hanginghatstudios.com/pigasus-faq/#acc-tb_obg1300-0
https://forum.skybox.xyz/d/157-filename-rules-for-vr-format

It may be nice to have a field to input/select a filename suffix.

>resume

I think it is possible if the following two conditions are true.
1. Exact same input settings
2. Temporary video file (`_tmp_*.mp4`) is not corrupted (If user cancels the operation, it is not corrupt, but if the program terminates abnormally, it may be corrupt)

>Anaglyph

This can be implemented, but I don't see any users actually using it.

>your own depth maps

If a user has the motivation to draw depth maps manually, I think they should use another software, such as Blender.
It is more flexible, with displacement modifier, sculpting, and authentic stereo camera settings available.

**Reply by gituser123456789000** [reply]
_LR_Flat worked to automatically display content correctly with DeoVR Player. Content with the original output name ending in _LRF would display at VR180 if that was my most previous or default setting and all of the flat content would have to be changed manually before being able to view it correctly

Negative Strength values, I agree, turned out to be useless in this program. The content could appear sunken in to the frame, but flat. After a lot of testing by multiple people, we pretty much agree the default of 2.0 works best with most content, and fooling around with testing outside of that value doesn't stray too far (usually only as far away as 1.0 or 3.0), but there could be some content that needs higher Strength to get the desired effect. We unlocked the custom allowed values to -20 through 20 just in case there is ever a use case for this. It doesn't hurt to allow the option, even if the recommended values in the drop-down menu would mostly be used. Same with Convergence Plane. I always use 0.5, but allowing -20 to 20 doesn't hurt anyone.. and it avoids people having to think 'what if'.. what if the program allowed me to try higher or lower values for some settings. Now they would be able to test for themselves and be sure they're getting the best result possible.

Edge Fix, someone did find that 5 could be useful in some content they were creating, so unlocking Edge Fix limitations would be good also. I personally currently don't use it or make it a very low value, but we unlocked it up to 8 and even tested up to 50 at one point, just to exaggerate the effect to understand exactly what it was doing. 8 or 10 I'm sure is plenty, but also unlocking this to 20 won't hurt anybody.

The ability to add custom Depth Resolution values is the biggest quick upgrade I see for the program.

Anaglyph I would personally put to use. I'm already happy with the program, but I'd definitely be happier with an Anaglyph option. There are people who like it and make content with it. It would be good to have as an option in this program and not have to go elsewhere when this program could do it at higher quality and faster with all of the customization options it allows for.

**Reply by nagadomi** [reply]
Regarding UI,
I would like to prevent novice users from entering unreasonable values.
However, I understand the request to enter arbitrary values for more advanced use.
So I will change input controls to satisfy both requests. At the moment, I plan to use Slider + TextBox.

>DeoVR Player

I will add this to my test env.

>Anaglyph

Ok. I bought anaglyph glasses on amazon. I hope I don't receive a defective product because according to the reviews some of them are bad quality or don't work.
For personal use, I think Google Cardboard is better, but I understand that Anaglyph has some advantages.

**Reply by gituser123456789000** [reply]
https://www.amazon.com/Plastic-3D-Glasses-Original-Anaglyph/dp/B00IIANEIY/

I got these ^

You get the glasses type and a couple of the cheaper cardboard type, so you can compare and see which you prefer. I use the glasses. I searched around and found those to be the best reviewed. Others have complaints that the colors aren't correct or that the glasses broke easily. These didn't have complaints about the color of the lenses. It comes with that book of some anaglyph images also. Not bad to start with. You may or may not enjoy it. I looked at it for a couple minutes and that was that. Looking at what you create or find on PC is where it's better. VR is obviously better than anaglyph, but for such a low price, anaglyph glasses are nice to have and check out content every once in a while

**Reply by gituser123456789000** [reply]
As we're learning more and more about the program and getting more experienced with it, the results are getting better and better. I'm now becoming a fan of Edge Fix. I was previously worried that the edges would look ghostly and smeared, but it's better than the alternative of not using it, which could leave sharp, much more distracting pixelation or artifacts. If Edge Fix does result in a ghostly smeared look of the edges that also distorts the background in general, but most noticeable with straight lines in the background... we found that Foreground Scale might clean that up (depending on the content).. but it often doesn't need it. I was able to push way too much Strength, at 6.0, which caused very poor edges on its own, but then with Edge Fix (I used 8 in this case) and Foreground Scale (3 in this case), the edges became beautiful, crisp, clean. Maybe not absolute perfection, but for that high of Strength... and I could have played with the values more to further improve it... wow. Impressive.

**Reply by nagadomi** [reply]
artifacts and distortions occur where there is a large depth difference between neighboring pixels.
For example, the boundary(edge) between the sky and people.
Edge fix has the effect of shifting the distorted area from the foreground to the background.
If people's faces are distorted, it is not good, but if the sky is distorted, it does not matter

**Reply by gituser123456789000** [reply]
You're doing excellent work, nagadomi. We see the Commits. I'm excited to check out what you've done with Anaglyph. I see you really took the time to research into it and provided different options, like Half-color, which reduces or removes ghosting.. and the other optimized versions will be very interesting to try. Check the Dubois method. This is supposed to create better results compared to standard anaglyph.. I'm not sure if it's the same or very similar to the optimized versions you've already added or not.

This is great that you went beyond just adding 1 bare basic red/cyan anaglyph option. I think the additional options you added will produce better results.

Also, have a look into yellow/blue anaglyph options. This is supposed to produce better color reproduction compared to red/cyan, but it's much less commonly used, since I suppose it came decades after red/cyan. I haven't tried it yet, personally, but if your program had the option to make them, I'd start being on the lookout for yellow/blue anaglyph glasses.

I also had a thought about adding frame interpolation.. the ability to increase the frame rates.

Not immediately necessary.. there are other programs to do this, such as Flowframes, but again, having everything all in your program would make the process a lot easier.

**Reply by nagadomi** [reply]
I checked a little about the Dubois method, but have not yet implemented it due to some uncertainties.
1. Some major implementations do not perform gamma correction. It is a incorrect implementation, but if it is popular with users, then it may not be wrong.
2. The original parameters is optimized for CRT. We are using LCD.

EDIT 3.:
In the [author's own documentation](https://www.site.uottawa.ca/~edubois/anaglyph/LeastSquaresHowToPhotoshop.pdf), it is described as "The first and second terms should be clipped to [0,1] before adding".
However, in that case, some values of matrix will be meaningless.

I'll try it anyway and add it if it's not too bad.

For other colored glasses, support itself is easy, but it should start with the buying of the glass.
Since red-cyan seems to be much more popular than the others, so, I supported only that for now.

**Reply by nagadomi** [reply]
I added Dubois Method to dev branch.
Not sure if this is right implementation, but it looks better than the other methods.

**Reply by gituser123456789000** [reply]
Absolutely fantastic work. I just updated and see a bunch of improvements.. added features, unlocked limitations

It looks like you can enter custom Depth Resolution values, custom Edge Fix values, new option for Pixel Format, custom Max FPS values, custom CRF values... a bunch of Anaglyph options in the Stereo Format section

**Reply by gituser123456789000** [reply]
My opinion on the anaglyph models... Ranking: (may vary in certain photos, but in general..)
Top Tier.. these have potential to look the best or near it
**wimmer** - Often looks the best in terms of eliminating ghosting, without having an unusual color tint to it.. may be easiest on the eyes.. it's like halfcolor and a little extra of what halfcolor fixes from fullcolor. It's sometimes or often dark though.
**wimmer2** - if wimmer is too dark, wimmer 2 may look best. It brightens things up, but has a pinkish/light red tint to it
**halfcolor** - wimmer does slightly better than what halfcolor does with removing ghosting.. halfcolor can still have a minor near-ghosting look that wimmer clears up completely.. halfcolor does retain slightly more original color compared to the wimmers though
**fullcolor** - can be fine if the image doesn't have colors that cause ghosting.. but most images have colors that cause ghosting. If there is large amounts of ghosting, this model is the worst.. but if there are zero colors in the original image that cause ghosting, this could potentially look the best since it retains the most original color

Bottom tier.. it looks like these two never have potential to be the best looking anaglyph model
**dubois** - skin tones and certain colors get turned to yellow. People look near Simpsons characters. May fix all ghosting, but doesn't look great in my opinion (I see you noted that it may not be correctly implemented)
**grayscale** - pretty much not worth it unless you love black and white.. even the yellow tinted dubois that still has some colors is better than no color at all

**Reply by gituser123456789000** [reply]
I saw in utils.py about the suffix redcyan being "# temporary"

That's something I noticed in testing.. needing to add custom suffixes to the end of the images so I could put the in the same folder and compare the anaglyph models side by side, back and forth

So it looks like you're already aware of this, but just in case, it would be best if they each had custom suffix names..
_redcyan_wimmer
_redcyan_wimmer2
_redcyan_halfcolor
_redcyan_color
_redcyan_dubois
_redcyan_gray

**Reply by nagadomi** [reply]
>dubois

It is normal behavior for the color to be yellow. I think this is `redcyan` limitation rather than `dubois`.
I have confirmed that the color is roughly the same as the result of the author's own conversion.
redcyan: https://www.flickr.com/photos/e_dubois/4576133130/
src: https://www.flickr.com/photos/e_dubois/4575500345/
I think the good point about `dubois` is that there is less discomfort between the left and right eye and less ghosting.
In the other methods, the colors seen by the left and right eyes seem to be very different.
Anaglyph visibility depends on the screen, glasses, and personal eye properties, so I think it is not easy to evaluate subjectively, though.

`wimmer` does not use the red channel of the source image at all. It generates the left and right images from blue and green channel only. wimmer2 uses a little red channel.

>suffix

I will change it for each method.

**Reply by nagadomi** [reply]
I added `dubois2`. This version ignores "The first and second terms should be clipped to [0,1] before adding" from spec. This one looks more closely to the author's result.
Also added anaglyph method name to the output filename.

**Comment by loawizard** [comment]
Got it!

Thank you for all your wonderful work.

**Comment by SaibotC** [comment]
Hi @loawizard, could you share the discord link again? It is expired now. Thanks

**Reply by loawizard** [reply]
https://discord.gg/xupjcXNA

**Reply by zelenukhin** [reply]
> https://discord.gg/xupjcXNA

Unfortunately the link has expired. May I ask you to share the actual link.

**Reply by loawizard** [reply]
https://discord.gg/SDCfvupZ

**Reply by dimdieo** [reply]
link has expired

**Reply by loawizard** [reply]
> link has expired

https://discord.gg/V7zkDv7Y

**Reply by gituser123456789000** [reply]
My edits keep advertising for this other server I'm not in 🤦

**Reply by loawizard** [reply]
> My edits keep advertising for this other server I'm not in 🤦

What you mean?

**Reply by thevonfamily89-wq** [reply]
@loawizard May I have a working invite? the links are expired thanks

**Comment by thevonfamily89-wq** [comment]
@loawizard May I have a working invite? the links are expired

**Reply by AIVFI** [reply]
> @loawizard May I have a working invite? the links are expired

https://discord.gg/yheQKEY2

**Reply by blu31c33** [reply]
id love an invite too. thanks!

**Reply by diverswan9** [reply]
> id love an invite too. thanks!

https://discord.gg/7EjpGCVK

**Reply by blu31c33** [reply]
i keep missing out on the discord invite.  can i please get one?  thanks!

**Reply by diverswan9** [reply]
https://discord.gg/HNYZQe9C

**Comment by MASTABA1965ARV** [comment]
The link is marked as expired, can I get one? Thanks!

**Reply by diverswan9** [reply]
https://discord.gg/8fqkHRDGx

**Comment by rsslcs** [comment]
@diverswan9 @loawizard 
Old link expired, could I please have a new one? Thanks!

**Reply by diverswan9** [reply]
https://discord.gg/h9FjFZgJ9
